### PR TITLE
[BW-1126] Workflow-level stdout / stderr expressions fail, hang execution forever

### PR DESCRIPTION
**Pull request by kpierre13**

### CODE REVIEW

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala [hunk 1]
```diff
@@ -26,26 +27,91 @@ object AstToWorkflowDefinitionElement {
...
 
-    (inputsSectionValidation, outputsSectionValidation, metaSectionValidation, parameterMetaSectionValidation) mapN {
-      (validInputs, validOutputs, meta, parameterMeta) =>
+    (inputsSectionValidation, outputsSectionValidation, metaSectionValidation, parameterMetaSectionValidation, intermediateValueDeclarationStdoutCheck, intermediateValueDeclarationStderrCheck) mapN {
+      (validInputs, validOutputs, meta, parameterMeta, _, _) =>
       WorkflowDefinitionElement(name, validInputs, graphSections.toSet, validOutputs, meta, parameterMeta, sourceLocation)
     }
   }
 
+  private def checkIfStdoutInputExist(inputSection: Option[InputsSectionElement]): ErrorOr[Option[String]] = {
```
Comment (aednichols) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala:56]: I only have a few minutes here before a meeting but I wanted to say this code looks good!

If there's one thing we could improve, I think we may be able to use the superclass `WorkflowBodyElement` (and equivalent superclasses for subsequent element types) to de-duplicate some of the `check...` functions.

**Review by cjllanwarne — APPROVED** [review]
I have a few high level thoughts and comments, happy to mob to go over them. But the top level comment is that this is well coded and correct, works as expected, and delivers value. So, I have no problem with merging as is.

Some comments and thoughts for improving this either now or in the future, in the order I would probably think about doing them. Happy to mob with you on any of these as a learning exercise, and equally happy to let them go if you'd rather commit some value here and move on to other things!

1. I think we could probably use unit tests to perform many of the checks currently in centaur tests. I think they would run in significantly less time and be significantly less flaky, just by being unit rather than integration tests. That would give us better feedback and faster turnaround times maintaining this now and in the future.
2.  I agree with Adam that some of the functions look like they could be combined or generalized somehow to reduce the amount of similar-looking code (and as a second plug - having good - and quick - unit tests to validate changes as you make them will be very beneficial if you attempt that refactor)
3. We are now covered if people use a single `stdout()` as their output expression, but are still vulnerable to calls which are nested inside other expressions, like `Array[File] outAndError = [ stdout(), stderr() ]`. I can't think of any particular use cases for a situation like that, but in _theory_ it's still possible.

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala (hunk 1 above)
Comment (salonishah11) [reply wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala:56]: +1 to Adam's suggestion. I think we can also use super class `WorkflowGraphElement` which is what `IntermediateValueDeclarationElement`, `OutputDeclarationElement` and `InputDeclarationElement` extend. We can also use generics for `StdoutElement` and `StderrElement` so that we don't need two different functions for this. I was thinking of something on the lines of
```
private def check[A <: WorkflowGraphElement, B](section: Seq[A], expressionType: B, expressionName: String): ErrorOr[Option[String]] = {
    if (section.map(_.expression).exists(_.isInstanceOf[expressionType.type])) {
      s"Workflow cannot have $expressionName expression at workflow-level.".invalidNel
    } else None.validNel
  }
```
But currently it throws error saying `value expression is not a member of type parameter A` since `WorkflowGraphElement` is a very broad super class and doesn't have the parameter `expression` in its class. I don't have a fix for this yet, but I will look into this more on Monday.

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala [hunk 2]
```diff
@@ -26,26 +27,91 @@ object AstToWorkflowDefinitionElement {
...
+    } else None.validNel
+  }
+
   private def validateSize[A](elements: Vector[A], sectionName: String, numExpected: Int): ErrorOr[Option[A]] = {
     val sectionValidation: ErrorOr[Option[A]] = if (elements.size > numExpected) {
       s"Workflow cannot have more than $numExpected $sectionName sections, found ${elements.size}.".invalidNel
     } else {
       elements.headOption.validNel
     }
+
```
Comment (salonishah11) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala:114]: Revert this change

File: wdl/transforms/draft3/src/test/scala/AstToWorkflowDefinitionElementSpec.scala [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import wom.types.{WomSingleFileType, WomStringType}
+import wdl.transforms.base.ast2wdlom.AstToWorkflowDefinitionElement._
+
+
+class AstToWorkflowDefinitionElementSpec extends AnyFlatSpec{
+  behavior of "Check Stdouts and Stderrs"
+
+  val mockInputSectionStdout = Option(InputsSectionElement(Vector(InputDeclarationElement(PrimitiveTypeElement(WomSingleFileType), "i", Some(StdoutElement)))))
+  val mockInputSectionStderr= Option(InputsSectionElement(Vector(InputDeclarationElement(PrimitiveTypeElement(WomSingleFileType), "i", Some(StderrElement)))))
+  val mockInputSectionNonStd= Option(InputsSectionElement(Vector(InputDeclarationElement(PrimitiveTypeElement(WomStringType), "more", Some(StringLiteral("more"))))))
```
Comment (cjllanwarne) [inline wdl/transforms/draft3/src/test/scala/AstToWorkflowDefinitionElementSpec.scala:15]: For consistent spacing
```suggestion
  val mockInputSectionStderr = Option(InputsSectionElement(Vector(InputDeclarationElement(PrimitiveTypeElement(WomSingleFileType), "i", Some(StderrElement)))))
  val mockInputSectionNonStd = Option(InputsSectionElement(Vector(InputDeclarationElement(PrimitiveTypeElement(WomStringType), "more", Some(StringLiteral("more"))))))
```

File: wdl/transforms/draft3/src/test/scala/AstToWorkflowDefinitionElementSpec.scala [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+  it should "return an error when there is an stderr in input section" in {
+    val testInputs = checkIfStdInputExist(mockInputSectionStderr, StderrElement, "stderr")
+    testInputs match {
+      case Valid(_) => fail("Input section contained stderr. Should have failed.")
+      case Invalid(e) => e.head should be("Workflow cannot have stderr expression in input section at workflow-level.")
+    }
+  }
+
+  it should "not throw an error for non-stdout/stderr in the inputs section" in {
```
Comment (cjllanwarne) [inline wdl/transforms/draft3/src/test/scala/AstToWorkflowDefinitionElementSpec.scala:43]: One nice thing about the validations is that they don't ever _throw_ things, the error is wrapped up in the return type to be handled whenever you want to handle it.

```suggestion
  it should "not return an error for non-stdout/stderr in the inputs section" in {
```

File: wdl/transforms/draft3/src/test/scala/AstToWorkflowDefinitionElementSpec.scala [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  it should "return an error when there is an stderr in output section" in {
+    val testOutputs = checkIfStdOutputExists(mockOutputSectionStderr, StderrElement, "stderr")
+    testOutputs match {
+      case Valid(_) => fail("Output section contained stderr. Should have failed.")
+      case Invalid(e) => e.head should be("Workflow cannot have stderr expression in output section at workflow-level.")
+    }
+  }
+  it should "not throw an error for non-stdout/stderr in the outputs section" in {
```
Comment (cjllanwarne) [inline wdl/transforms/draft3/src/test/scala/AstToWorkflowDefinitionElementSpec.scala:66]:
```suggestion
  it should "not return an error for non-stdout/stderr in the outputs section" in {
```

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala [hunk 6]
```diff
@@ -26,26 +27,65 @@ object AstToWorkflowDefinitionElement {
...
+    inputSection match {
+      case Some(section) =>
+        if (section.inputDeclarations.flatMap(_.expression).exists(_.isInstanceOf[expressionType.type])) {
+          s"Workflow cannot have $expressionName expression in input section at workflow-level.".invalidNel
+        } else None.validNel
+      case None => None.validNel
+    }
+  }
+
+  def checkIfStdOutputExists[A](outputSection: Option[OutputsSectionElement], expressionType: FunctionCallElement, expressionName: String): ErrorOr[Option[String]] = {
```
Comment (cjllanwarne) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala:66]: What's the `[A]` doing?

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala [hunk 7]
```diff
@@ -26,26 +27,65 @@ object AstToWorkflowDefinitionElement {
...
+    outputSection match {
+      case Some(section) =>
+        if (section.outputs.map(_.expression).exists(_.isInstanceOf[expressionType.type])) {
+          s"Workflow cannot have $expressionName expression in output section at workflow-level.".invalidNel
+        } else None.validNel
+      case None => None.validNel
+    }
+  }
+
+  def checkStdIntermediates[A](intermediate: Vector[IntermediateValueDeclarationElement], expressionType: FunctionCallElement, expressionName: String): ErrorOr[Option[String]] = {
```
Comment (cjllanwarne) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala:76]: Going one step further, these three functions are similar enough that I think you could probably (if so inclined) combine them into a single function like

```
def checkForExpressionTypes[A](
    elementsToCheck: Iterable[A],
    expressionFinder: A => Iterable[ExpressionElement],
    expressionTypesToFind: Iterable[FunctionCallElement], 
    expressionName: String) = {
  ???
}
```

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala [hunk 8]
```diff
@@ -26,26 +27,65 @@ object AstToWorkflowDefinitionElement {
...
 
-    (inputsSectionValidation, outputsSectionValidation, metaSectionValidation, parameterMetaSectionValidation) mapN {
-      (validInputs, validOutputs, meta, parameterMeta) =>
+    (inputsSectionValidation, outputsSectionValidation, metaSectionValidation, parameterMetaSectionValidation, intermediateValueDeclarationStdoutCheck, intermediateValueDeclarationStderrCheck) mapN {
+      (validInputs, validOutputs, meta, parameterMeta, _, _) =>
       WorkflowDefinitionElement(name, validInputs, graphSections.toSet, validOutputs, meta, parameterMeta, sourceLocation)
     }
   }
 
+  def checkIfStdInputExist(inputSection: Option[InputsSectionElement], expressionType: FunctionCallElement, expressionName: String): ErrorOr[Option[String]] = {
```
[the remark this answers is not in the document]
Comment (kpierre13) [reply wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala:56]: When talking about 'read the Option[String]', are you talking about reading the error message that gets generated when running an stdout/stderr wdl?

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala (hunk 7 above)
Comment (kpierre13) [reply wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala:76]: I chose to have them all separate in case another noob like me comes along and is working on this, but implementing it this way would be cleaner I suppose. I'm torn between the two!

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (cjllanwarne) [reply wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala:56]: If you unpack it, the `ErrorOr[A]` type is expanded behind the scenes into a `Validated[A, NonEmptyList[String]]`. That `Validated` type can either be an "A" (if it succeeds) OR a list of error message strings if it fails. So the `ErrorOr` type only needs to include the "successful" type, the "failure" type is already encoded into it

**Review by cjllanwarne — APPROVED** [review]
Other than the ErrorOr return type, my comments are mostly "if you feel like it" level

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala (hunk 6 above)
Comment (kpierre13) [reply wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala:66]: Thank you for catching this. It was an artifact of an older function.

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala [hunk 9]
```diff
@@ -26,26 +27,65 @@ object AstToWorkflowDefinitionElement {
                               sourceLocation: Option[SourceFileLocation],
                               bodyElements: Vector[WorkflowBodyElement]) = {
 
-    val inputsSectionValidation: ErrorOr[Option[InputsSectionElement]] = validateSize(bodyElements.filterByType[InputsSectionElement], "inputs", 1)
-    val outputsSectionValidation: ErrorOr[Option[OutputsSectionElement]] = validateSize(bodyElements.filterByType[OutputsSectionElement], "outputs", 1)
+    val inputsSectionValidation: ErrorOr[Option[InputsSectionElement]] = for {
+      inputValidateElement <- validateSize(bodyElements.filterByType[InputsSectionElement], "inputs", 1): ErrorOr[Option[InputsSectionElement]]
+      _ <- checkIfStdInputExist(inputValidateElement, StdoutElement, "stdout")
+      _ <- checkIfStdInputExist(inputValidateElement, StderrElement, "stderr")
```
Comment (aednichols) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/ast2wdlom/AstToWorkflowDefinitionElement.scala:33]: I would love to rename this function to a more generic `checkDisallowedElement` instead of including `Std...` in the name.

Looking at this call site alone (without viewing the function definition) it is not clear whether we are checking for `"stdout"` because we disallow it, or because we require it, or we're just curious ...

( applies here and elsewhere )

**Review by aednichols — APPROVED** [review]

**Comment by kpierre13** [comment]
Closing due to PR moving elsewhere.
